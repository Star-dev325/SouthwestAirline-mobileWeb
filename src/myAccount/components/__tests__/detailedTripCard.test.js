jest.mock('@swa-ui/encryption', () => ({
  useHref: jest.fn().mockReturnValue({ href: 'mock_href' })
}));

import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import DetailedTripCard from 'src/myAccount/components/detailedTripCard';
import UpcomingTripBuilder from 'test/builders/apiResponse/v1/mobile-misc/page/upcoming-trips/upcomingTripBuilder';
import UpcomingTripsBuilder from 'test/builders/apiResponse/v1/mobile-misc/page/upcoming-trips/upcomingTripsBuilder';
import createMockStore from 'test/unit/helpers/createMockStore';

const mockStore = createMockStore();

describe('DetailedTripCard', () => {
  describe('when trip is within 24', () => {
    it('should show the boarding banner when trip has a segment that is boarding', () => {
      const page = new UpcomingTripBuilder().withBannerTypeAndText('POSITIVE', 'Now Boarding').build();
      const { _links, confirmationNumber, dates, pages, tripType } = page;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('.banner_positive .banner-name').textContent).toContain('Now Boarding');
    });
  });

  describe('when there is a cancelled flight', () => {
    it('should show red banner', () => {
      const upcomingTripWithCancelledFight = new UpcomingTripsBuilder().withCancelledFlight().build();
      const oneSegmentEligibleTrip = upcomingTripWithCancelledFight.upcomingTripsPage[0];
      const { _links, confirmationNumber, dates, pages, tripType } = oneSegmentEligibleTrip;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('.banner_negative .banner-name').textContent).toContain('Cancelled');
      expect(container.querySelector('.banner_negative .boarding-time').textContent).toContain('Boards 05:15AM');
    });
  });

  describe('segment header', () => {
    it('should show the confirmation number', () => {
      const tripThatIsCheckInEligibleTwoSegments = new UpcomingTripBuilder().with2SegmentsOnBound().build();
      const { _links, confirmationNumber, dates, pages, tripType } = tripThatIsCheckInEligibleTwoSegments;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelectorAll('.trip-card-header--confirmation-number_number')).toHaveLength(2);
    });

    it('should display the trip card header with the destination first and a horizontal rule beneath the trip card header', () => {
      const upcomingTripThatIsWithin24HoursApiResponse = new UpcomingTripsBuilder()
        .withViewBoardingPositionFlight()
        .build();
      const trip = upcomingTripThatIsWithin24HoursApiResponse.upcomingTripsPage[0];
      const { _links, confirmationNumber, dates, pages, tripType } = trip;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container).toMatchSnapshot();
    });

    describe('header uses segment destination description not trip destination description', () => {
      it('should show trip destination description for 1st trip card when not departed', () => {
        const upcomingNotDepartedTrip = new UpcomingTripsBuilder().withOneWayFlightHasConnection().build()
          .upcomingTripsPage[0];
        const { _links, confirmationNumber, dates, pages, tripType } = upcomingNotDepartedTrip;

        const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });
        const expectedDestinationAirport = upcomingNotDepartedTrip.pages[0].destinationDescription;

        expect(
          container.querySelector('.trip-card-header').querySelector('.trip-card-header--destination-airport')
            .textContent
        ).toContain(expectedDestinationAirport);
      });

      it('should show trip destination description for 2nd trip card when not departed', () => {
        const upcomingNotDepartedTrip = new UpcomingTripsBuilder().withOneWayFlightHasConnection().build()
          .upcomingTripsPage[0];
        const { _links, confirmationNumber, dates, pages, tripType } = upcomingNotDepartedTrip;

        const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });
        const expectedDestinationAirport = upcomingNotDepartedTrip.pages[1].destinationDescription;

        expect(
          container.querySelectorAll('.trip-card-header')[1].querySelector('.trip-card-header--destination-airport')
            .textContent
        ).toContain(expectedDestinationAirport);
      });

      it('should show segment destination description for 1st trip card when a connection', () => {
        const upcomingDepartedTrip = new UpcomingTripsBuilder().withDepartedFlightHasConnection().build()
          .upcomingTripsPage[0];
        const { _links, confirmationNumber, dates, pages, tripType } = upcomingDepartedTrip;

        const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });
        const expectedDestinationAirport = upcomingDepartedTrip.pages[0].destinationDescription;

        expect(
          container.querySelector('.trip-card-header').querySelector('.trip-card-header--destination-airport')
            .textContent
        ).toContain(expectedDestinationAirport);
      });
    });
  });

  describe('standbyCard', () => {
    let tripThatIsCheckInEligibleTwoSegments;

    beforeEach(() => {
      tripThatIsCheckInEligibleTwoSegments = new UpcomingTripBuilder().build();

      tripThatIsCheckInEligibleTwoSegments.pages[0].standbyFlight = {
        arrivalAirportCode: 'AUS',
        arrivalTime: '09:55',
        departureTime: '09:00',
        flightNumber: '726',
        hasWifi: true,
        viewStandbyList: {
          body: {},
          href: '/v1/mobile-air-operations/page/standby',
          method: 'GET',
          query: {}
        }
      };
    });

    it('should show standby banner for non-rev passenger', () => {
      tripThatIsCheckInEligibleTwoSegments.pages[0].isNonRevPnr = true;
      tripThatIsCheckInEligibleTwoSegments.pages[0].bannerType = 'DEFAULT';
      tripThatIsCheckInEligibleTwoSegments.pages[0].bannerText = 'You are on standby to MDW';

      const { _links, confirmationNumber, dates, pages, tripType } = tripThatIsCheckInEligibleTwoSegments;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('.banner.banner_default .banner-name').textContent).toContain(
        'You are on standby to MDW'
      );
    });
  });

  describe('carousel', () => {
    it('should not have a carousel for one segment', () => {
      const oneSegmentEligibleTrip = new UpcomingTripBuilder().build();
      const { _links, confirmationNumber, dates, pages, tripType } = oneSegmentEligibleTrip;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('.carousel-dots-container')).toBeNull();
    });

    it('should have a carousel for multiple segments', () => {
      const tripThatIsCheckInEligibleTwoSegments = new UpcomingTripBuilder().with2SegmentsOnBound().build();
      const { _links, confirmationNumber, dates, pages, tripType } = tripThatIsCheckInEligibleTwoSegments;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelectorAll('.carousel-dots-container')).toHaveLength(1);
    });
  });

  describe('overnight', () => {
    it('should show overnight indicator when isOvernight is true', () => {
      const page = new UpcomingTripBuilder().withOvernight().build();
      const { _links, confirmationNumber, dates, pages, tripType } = page;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('[data-qa="overnight-indicator"]')).toMatchSnapshot();
    });

    it('should not show overnight indicator when isOvernight is false', () => {
      const page = new UpcomingTripBuilder().withoutOvernight().build();
      const { _links, confirmationNumber, dates, pages, tripType } = page;

      const { container } = createComponent({ _links, confirmationNumber, dates, pages, tripType });

      expect(container.querySelector('[data-qa="overnight-indicator"]')).toMatchSnapshot();
    });
  });

  const createComponent = (props) => {
    const defaultProps = {
      UPGRADED_BOARDING: false,
      onCheckInButtonClick: () => {},
      onClickDetailsButton: () => {},
      onClickEBCheckInButton: () => {},
      onClickStandbyList: () => {},
      onSelectNewFlightForCancelledFlight: () => {},
      onUpgradedBoardingButtonClick: () => {},
      onViewBoardingPassButtonClickCb: () => {},
      onViewBoardingPositionsButtonClick: () => {},
      pnr: {},
      tripIndex: 0
    };

    const tripCardProps = { ...defaultProps, ...props };
    const store = mockStore({
      app: {
        toggles: {
          AIRCRAFT_TYPE_TRIPCARD: false
        }
      }
    });

    return render(
      <Provider store={store}>
        <DetailedTripCard {...tripCardProps} />
      </Provider>
    );
  };
});
