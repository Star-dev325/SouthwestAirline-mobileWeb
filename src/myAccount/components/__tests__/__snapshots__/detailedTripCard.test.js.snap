// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedTripCard overnight should not show overnight indicator when isOvernight is false 1`] = `null`;

exports[`DetailedTripCard overnight should show overnight indicator when isOvernight is true 1`] = `
<div
  class="multi-day-indicator multi-day-indicator--overnight-red"
  data-qa="overnight-indicator"
>
  <i
    class="multi-day-indicator--icon multi-day-indicator--icon-small icon icon_next-day-indicator"
  />
  <p
    class="multi-day-indicator--font-small"
  >
    AIR_BOOKING__SHOPPING__OVERNIGHT
  </p>
</div>
`;

exports[`DetailedTripCard segment header should display the trip card header with the destination first and a horizontal rule beneath the trip card header 1`] = `
<div>
  <div>
    <div>
      <div
        class="detailed-trip-card"
      >
        <div
          class="trip-card-header"
        >
          <div
            class="trip-card-header--date-and-destination"
          >
            <span
              class="trip-card-header--destination-airport"
            >
              Newark
            </span>
            <span
              class="trip-card-header--trip-date"
            >
              Fri, Sep 29
            </span>
          </div>
          <div
            class="trip-card-header--icon-or-confirmation-number"
          >
            <span
              class="trip-card-header--confirmation-number"
            >
              <span
                class="trip-card-header--confirmation-number_label"
              >
                MY_ACCOUNT__CONFIRMATION_NUMBER
              </span>
              <span
                class="trip-card-header--confirmation-number_number"
              >
                QIQNWQ
              </span>
            </span>
          </div>
        </div>
        <hr
          class="detailed-trip-card--divider"
        />
        <div>
          <div
            class="flight-segment-details"
          >
            <div
              class="flight-segment-details--time-and-flight"
            >
              <div
                class="flight-segment-details--time-block"
              >
                <div
                  class="flight-segment-details--time-block-label"
                >
                  SHARED__FLIGHT_SEGMENT_DETAILS__DEPARTS
                </div>
                <div
                  class="flight-segment-details--time-block-time"
                >
                  6:00
                  <span
                    class="flight-segment-details--time-block-time-ampm"
                  >
                    AM
                  </span>
                </div>
              </div>
              <div
                class="flight-segment-details--flight-number"
              >
                <div>
                  SHARED__FLIGHT_SEGMENT_DETAILS__FLIGHT
                   
                  179
                </div>
                <i
                  class="icon icon_airplane"
                />
              </div>
              <div
                class="flight-segment-details--time-block align-right"
              >
                <div
                  class="flight-segment-details--time-block-label"
                >
                  SHARED__FLIGHT_SEGMENT_DETAILS__ARRIVES
                </div>
                <div
                  class="flight-segment-details--time-block-time"
                >
                  9:00
                  <span
                    class="flight-segment-details--time-block-time-ampm"
                  >
                    AM
                  </span>
                </div>
              </div>
            </div>
            <div
              class="flight-segment-details--airports"
            >
              <div
                class="flight-segment-details--airport-info pr3"
              >
                Chicago (Midway), IL - MDW
              </div>
              <div
                class="flight-segment-details--airport-info pl3 align-right"
              >
                New York/Newark, NJ - EWR
              </div>
            </div>
          </div>
          <div
            class="boarding-info boarding-information center pt5 pb5 mt3"
          >
            <div
              class="boarding-information--item-row"
            >
              <div
                class="boarding-information--item"
              >
                <div
                  class="boarding-information--item-label"
                >
                  SHARED__BOARDING_INFORMATION__GATE
                </div>
                <div
                  class="boarding-information--item-info"
                >
                  ---
                </div>
              </div>
              <div
                class="boarding-information--item"
              >
                <div
                  class="boarding-information--item-label"
                >
                  SHARED__BOARDING_INFORMATION__GROUP
                </div>
                <div
                  class="boarding-information--item-info"
                >
                  A
                </div>
              </div>
              <div
                class="boarding-information--item"
              >
                <div
                  class="boarding-information--item-label"
                >
                  SHARED__BOARDING_INFORMATION__POSITION
                </div>
                <div
                  class="boarding-information--item-info_right"
                >
                  20
                </div>
              </div>
            </div>
          </div>
          <div
            class="mt3"
          >
            <button
              class="detailed-trip-card--detail-button button"
              type="button"
            >
              View/Manage
            </button>
          </div>
          <div>
            <div
              class="mt3"
            >
              <button
                class="button button--fluid button--yellow"
                title="SHARED__BUTTON_TEXT__BOARDING_PASSES"
                type="button"
              >
                SHARED__BUTTON_TEXT__BOARDING_PASSES
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
